lib/translator/report_luns.cc

```diff
@@ -0,0 +1,111 @@
+uint32_t GetNsListLength(const nvme::IdentifyNamespaceList& ns_list) {
+  uint32_t size = 0;
+  for (uint32_t i = 0; i < nvme::kIdentifyNsListMaxLength; ++i) {
+    if (ns_list.ids[i] == 0) break;
+    ++size;
+  }
+  return size;
+}
+
+}  // namespace
+
+// Section 4.5
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode ReportLunsToNvme(Span<const uint8_t> scsi_cmd,
+                            nvme::GenericQueueEntryCmd& identify_cmd,
+                            Allocation& allocation, uint32_t& alloc_len) {
+  // Cast scsi_cmd to ReportLunsCommand
+  scsi::ReportLunsCommand rl_cmd;
+  if (!ReadValue(scsi_cmd, rl_cmd)) {
+    DebugLog("Malformed ReportLuns command");
+    return StatusCode::kInvalidInput;
+  }
+  if (rl_cmd.select_report != scsi::SelectReport::kRestrictedMethods &&
+      rl_cmd.select_report != scsi::SelectReport::kWellKnown &&
+      rl_cmd.select_report != scsi::SelectReport::kAllLogical) {
+    DebugLog("Invalid report luns select report %u", rl_cmd.select_report);
+    return StatusCode::kInvalidInput;
+  }
+
+  // Assign allocation length for downstream use
+  alloc_len = rl_cmd.alloc_length;
+
+  // Assign generic params to NVMe command
+  identify_cmd = {};
+  identify_cmd.opc = static_cast<uint8_t>(nvme::AdminOpcode::kIdentify);
+  identify_cmd.cdw[0] = 0x2;  // Set CNS to return namespace ID list
+
+  // Allocate prp & assign to command
+  if (allocation.SetPages(1, 0) == StatusCode::kFailure)
+    return StatusCode::kFailure;
+  identify_cmd.dptr.prp.prp1 = allocation.data_addr;
+  return StatusCode::kSuccess;
+}
+
+// Section 6.6
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode ReportLunsToScsi(const nvme::GenericQueueEntryCmd& identify_cmd,
+                            Span<uint8_t> buffer) {
+  if (buffer.size() < sizeof(scsi::ReportLunsParamData)) {
+    DebugLog("Insufficient buffer size");
+    return StatusCode::kFailure;
+  }
+
+  // Get ns_list and lun count from NVMe dptr
+  nvme::IdentifyNamespaceList* ns_list;
+  ns_list = reinterpret_cast<nvme::IdentifyNamespaceList*>(
+      identify_cmd.dptr.prp.prp1);
+  uint32_t lun_count = GetNsListLength(*ns_list);
+
+  // Define scsi response structure and determine attributes
+  uint32_t allocated_list_bytes =
+      buffer.size() - sizeof(scsi::ReportLunsParamData);
+  scsi::ReportLunsParamData rlpd = {};
+  rlpd.list_byte_length = lun_count * sizeof(scsi::LunAddress);
+  if (rlpd.list_byte_length > allocated_list_bytes) {
+    rlpd.list_byte_length = allocated_list_bytes -
+                            (allocated_list_bytes % sizeof(scsi::LunAddress));
+    lun_count = rlpd.list_byte_length / sizeof(scsi::LunAddress);
+  }
+
+  // Write response to buffer
+  if (!WriteValue(rlpd, buffer)) {
+    DebugLog("Buffer not large enough for report luns response header");
+    return StatusCode::kSuccess;
```

Comment:
why do we return success here?

Reply:
Truncating variable length data in SCSI is not meant to be an error